pyhf/pdf.py

```diff
@@ -131,6 +137,173 @@ def __init__(self, spec, **config_kwargs):
+                        mega_mods[s][m]['data']['mask']  += maskval
+                        mega_mods[s][m]['data']['uncrt'] += uncrt
+                    elif mtype == 'shapefactor':
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['mask'] += [maskval]*len(nom) #broadcasting
+                    else:
+                        raise RuntimeError
+            sample_dict = {
+                'name': 'mega_{}'.format(s),
+                'nom': mega_nom,
+                'modifiers': list(mega_mods[s].values())
+            }
+            mega_samples[s] = sample_dict
+        self.mega_samples = mega_samples
+        self.mega_mods    = mega_mods
+
+    def prep(self):
+        import numpy as np
+        self.normsys_histoset = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['lo'],
+                    [1.]*len(self.mega_samples[s]['nom']),
+                    self.mega_mods[s][m]['data']['hi'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'normsys' 
+        ])
+        self.normsys_mask = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['mask'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'normsys' 
+        ])
+        self.normsys_default = np.ones(self.normsys_mask.shape)
+
+
+        self.histosys_histoset = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['lo_data'],
+                    self.mega_samples[s]['nom'],
+                    self.mega_mods[s][m]['data']['hi_data'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'histosys' 
+        ])
+        self.histosys_mask = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['mask'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'histosys' 
+        ])
+        self.histosys_default = np.zeros(self.histosys_mask.shape)
+
+
+        self.normfactor_mask = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['mask'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'normfactor' 
+        ])
+        self.normfactor_default = np.ones(self.normfactor_mask.shape)
+        self.staterror_mask = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['mask'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'staterror' 
+        ])
+        self.staterror_default = np.ones(self.staterror_mask.shape)
+
```

Comment:
the default value for normfactor and staterror is a multiplication by 1.0 (a no-op) so we crate masks and the default values